nemo/collections/asr/parts/preprocessing/features.py

```diff
@@ -428,5 +476,175 @@ def forward(self, x, seq_len):
+                audio_len = audio_len + 2 * self.stft_pad_amount
+
+            spec_len = dali.fn.cast((audio_len // self.hop_length) + 1, dtype=dali.types.INT64)
+
+            # Additive gaussian noise (dither)
+            if self.training and self.dither > 0.0:
+                gaussian_noise = dali.fn.random.normal(device=device)
+                audio = audio + self.dither * gaussian_noise
+
+            # Preemphasis filter
+            if self.preemph > 0.0:
+                audio = dali.fn.preemphasis_filter(audio, preemph_coeff=self.preemph)
+
+            # Power spectrogram
+            spec = dali.fn.spectrogram(
+                audio,
+                nfft=self.n_fft,
+                power=self.mag_power,
+                window_fn=self.window_tensor_lst,
+                window_length=self.win_length,
+                window_step=self.hop_length,
+                center_windows=True,
+                reflect_padding=True,
+            )
+
+            # Spectrogram to Mel Spectrogram
+            spec = dali.fn.mel_filter_bank(
+                spec, sample_rate=self.sample_rate, nfilter=self.nfilt, freq_low=self.lowfreq, freq_high=self.highfreq,
+            )
+
+            # log features if required
+            if self.log:
+                eps = self.log_zero_guard_value_fn(torch.float32)
+                if self.log_zero_guard_type == "add":
+                    spec = spec + eps
+                elif self.log_zero_guard_type == "clamp":
+                    spec = dali.math.max(spec, eps)
+                else:
+                    raise ValueError("log_zero_guard_type was not understood")
+                # Natural Logarithm
+                spec = dali.fn.to_decibels(spec, multiplier=math.log(10), reference=1.0, cutoff_db=-120)
+
+            if self.frame_splicing > 1:
+                spec = self._daligraph_splice_frames(spec, self.nfilt, spec_len, stacking=self.frame_splicing)
+
+            # Trimming Spectrogram to match the reference implementation
+            start = dali.types.Constant(0, shape=[], dtype=dali.types.INT64, device='cpu')
+            spec = dali.fn.slice(spec, start, spec_len, axes=(1,))
+
+            # Normalization
+            normalization_axes = None
+            if self.normalize:
+                if self.normalize == "per_feature":
+                    normalization_axes = [1]
+                elif self.normalize == "all_features":
+                    normalization_axes = [0, 1]
+                elif "fixed_mean" in self.normalize and "fixed_std" in self.normalize:
+                    raise ValueError("Normalization with fixed mean/stddev not yet supported.")
+                    # TODO: implement
+                else:
+                    raise ValueError(f"Unknown normalization type: {self.normalize}")
+
+                # Normalization
+                spec = dali.fn.normalize(spec, axes=normalization_axes, epsilon=1e-5 ** 2, ddof=1)
```

Comment:
1e-10 eps is too low (maybe) for fp16? Or is this all processed under fp32 first and then finally cast to fp16?

Reply:
The whole pipeline works with fp32. We can cast it to float16 inside DALI, if needed, at the end of the pipeline, but the current pipeline doesn't do it.

Reply:
Ok sounds good.